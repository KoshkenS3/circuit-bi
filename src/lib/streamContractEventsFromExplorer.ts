import { logger } from "../utils/logger";
import _ERC20Abi from "../../data/interfaces/standard/ERC20.json";
import _BeefyVaultV6Abi from "../../data/interfaces/beefy/BeefyVaultV6/BeefyVaultV6.json";
import { ethers } from "ethers";
import { ERC20EventData } from "../lib/csv-transfer-events";
import * as lodash from "lodash";
import { Chain } from "../types/chain";
import { callLockProtectedExplorerUrl } from "./shared-resources/shared-explorer";
import { callLockProtectedRpc } from "./shared-resources/shared-rpc";
import { fetchContractCreationInfos } from "./fetch-if-not-found-locally";

const ERC20Abi = _ERC20Abi as any as JsonAbi;
const BeefyVaultV6Abi = _BeefyVaultV6Abi as any as JsonAbi;

type JsonAbi = {
  inputs?: {
    internalType: string;
    name: string;
    type: string;
    indexed?: boolean;
  }[];
  stateMutability?: string;
  type: string;
  anonymous?: boolean;
  name?: string;
  outputs?: {
    internalType: string;
    name: string;
    type: string;
  }[];
}[];

interface ExplorerLog {
  address: string;
  topics: (string | null)[];
  data: string;
  blockNumber: number | string;
  timeStamp: number | string;
  gasPrice: number | string;
  gasUsed: number | string;
  logIndex: number | string;
  transactionHash: string;
  transactionIndex: number | string;
}

// be nice to explorers or you'll get banned
async function fetchExplorerLogsPage<TRes extends { blockNumber: number }>(
  chain: Chain,
  contractAddress: string,
  abi: JsonAbi,
  eventName: string,
  fromBlock: number,
  toBlock: number | null,
  formatEvent: (event: ExplorerLog) => TRes,
  fromAddress?: string
) {
  logger.debug(
    `[ERC20.T.EX] Fetching ${eventName} events from ${fromBlock} for ${chain}:${contractAddress}:${fromAddress}`
  );
  const eventTopic = getEventTopicFromJsonAbi(abi, eventName);
  const params: Record<string, string> = {
    module: "logs",
    action: "getLogs",
    address: contractAddress,
    topic0: eventTopic,
    fromBlock: fromBlock.toString(),
  };
  let didLimitBlockCount = false;
  if (toBlock) {
    params.toBlock = toBlock.toString();
    const blockCount = toBlock - fromBlock;
    if (blockCount > 10000) {
      logger.info(
        `[ERC20.T.EX] Limiting block count to 10k blocks to avoid 504s from explorers`
      );
      params.toBlock = (fromBlock + 10000).toString();
      didLimitBlockCount = true;
    }
  }
  if (fromAddress) {
    params.topic1 =
      "0x000000000000000000000000" + fromAddress.slice(2) /** remove "0x" */;
  }
  if (chain === "metis") {
    // https://andromeda-explorer.metis.io/api-docs
    // Error calling explorer https://andromeda-explorer.metis.io/api: {"message":"Required query parameters missing: topic0_1_opr","result":null,"status":"0"}
    // we use "or" because there is no topic 1
    params.topic0_1_opr = "or";
  }
  const rawLogs = await callLockProtectedExplorerUrl<ExplorerLog[]>(
    chain,
    params
  );
  let logs = rawLogs.map(formatEvent);
  const mayHaveMore = didLimitBlockCount ? true : logs.length === 1000;

  logger.verbose(
    `[ERC20.T.EX] Got ${logs.length} ${eventName} events for ${chain}:${contractAddress}:${fromAddress}, mayHaveMore: ${mayHaveMore}`
  );
  // remove last block data as the explorer may have truncated results
  // in the middle of a block
  if (mayHaveMore) {
    const lastLogBlock = logs[logs.length - 1];
    for (let i = logs.length - 1; i >= 0; i--) {
      if (logs[i].blockNumber === lastLogBlock.blockNumber) {
        logs.pop();
      } else {
        break;
      }
    }
  }
  return { logs, mayHaveMore };
}

const getEventTopicFromJsonAbi = lodash.memoize(
  function _getEventTopicFromJsonAbi(abi: JsonAbi, eventName: string) {
    const eventTypes = getEventTypesFromJsonAbi(abi, eventName);
    return ethers.utils.keccak256(
      ethers.utils.toUtf8Bytes(`${eventName}(${eventTypes.join(",")})`)
    );
  }
);

const getEventTypesFromJsonAbi = lodash.memoize(
  function _getEventTypesFromJsonAbi(abi: JsonAbi, eventName: string) {
    const eventConfig = abi.find(
      (abi) => abi.name === eventName && abi.type === "event"
    );
    if (!eventConfig || !eventConfig.inputs) {
      throw new Error(`${eventName} not found in abi`);
    }
    return eventConfig.inputs.map((input) => input.type);
  }
);

export function explorerLogToERC20TransferEvent(
  event: ExplorerLog
): ERC20EventData {
  const blockNumber = parseInt(
    ethers.BigNumber.from(event.blockNumber).toString()
  );
  const data =
    "0x" +
    event.topics
      .slice(1)
      .concat([event.data])
      .map((hexData: string | null) => (hexData ? hexData.slice(2) : hexData))
      .join("");
  const [from, to, value] = ethers.utils.defaultAbiCoder.decode(
    getEventTypesFromJsonAbi(ERC20Abi, "Transfer"),
    data
  );
  const datetime = new Date(
    ethers.BigNumber.from(event.timeStamp).toNumber() * 1000
  );
  return {
    blockNumber,
    datetime,
    from,
    to,
    value: value.toString(),
  };
}

export async function* streamERC20TransferEventsFromExplorer(
  chain: Chain,
  contractAddress: string,
  fromBlock: number,
  toBlock: number | null,
  fromAddress?: string
) {
  let mayHaveMore = true;
  while (mayHaveMore) {
    const pageRes = await fetchExplorerLogsPage(
      chain,
      contractAddress,
      ERC20Abi,
      "Transfer",
      fromBlock,
      toBlock,
      explorerLogToERC20TransferEvent,
      fromAddress
    );
    if (pageRes.logs.length === 0) {
      return;
    }
    yield* pageRes.logs;
    mayHaveMore = pageRes.mayHaveMore;
    fromBlock = pageRes.logs[pageRes.logs.length - 1].blockNumber + 1;
  }
}

interface BeefyVaultV6StrategyUpgradeEvent {
  blockNumber: number;
  datetime: Date;
  data: { implementation: string };
}

function explorerLogToBeefyVaultV6UpgradeStratEvent(
  event: ExplorerLog
): BeefyVaultV6StrategyUpgradeEvent {
  const blockNumber = parseInt(
    ethers.BigNumber.from(event.blockNumber).toString()
  );
  const data =
    "0x" +
    event.topics
      .slice(1)
      .concat([event.data])
      .map((hexData: string | null) => (hexData ? hexData.slice(2) : hexData))
      .join("");
  const [implementation] = ethers.utils.defaultAbiCoder.decode(
    getEventTypesFromJsonAbi(BeefyVaultV6Abi, "UpgradeStrat"),
    data
  );
  const datetime = new Date(
    ethers.BigNumber.from(event.timeStamp).toNumber() * 1000
  );
  return {
    blockNumber,
    datetime,
    data: {
      implementation,
    },
  };
}

export async function* streamBifiVaultUpgradeStratEventsFromExplorer(
  chain: Chain,
  contractAddress: string,
  startBlock: number,
  endBlock: number | null
) {
  const { blockNumber: deployBlockNumber, datetime: deployBlockDatetime } =
    await fetchContractCreationInfos(chain, contractAddress);

  // first, get all strategy upgrade events
  // it is very unlikely that there are more than 1000 events
  let allStrategyEvents: BeefyVaultV6StrategyUpgradeEvent[] = [];
  let mayHaveMore = true;
  let fromBlock =
    startBlock > deployBlockNumber ? startBlock : deployBlockNumber;
  while (mayHaveMore) {
    const pageRes = await fetchExplorerLogsPage(
      chain,
      contractAddress,
      BeefyVaultV6Abi,
      "UpgradeStrat",
      fromBlock,
      endBlock,
      explorerLogToBeefyVaultV6UpgradeStratEvent
    );
    allStrategyEvents = allStrategyEvents.concat(pageRes.logs);
    mayHaveMore = pageRes.mayHaveMore;
    if (pageRes.logs.length > 0) {
      fromBlock = pageRes.logs[pageRes.logs.length - 1].blockNumber + 1;
    } else {
      break;
    }
  }

  // now, we want to yield a fake event for the strategy on deploy
  // if the stragegy has never been upgraded, yield the current strategy at the deploy block time
  // if the stragegy has been upgraded, yield the strategy at the block preceeding the first upgrade block time

  let callOptions: { blockTag: number } | undefined = undefined;
  if (allStrategyEvents.length === 0) {
    callOptions = undefined;
  } else {
    const firstUpgradeBlockNumber = allStrategyEvents[0].blockNumber;
    callOptions = { blockTag: firstUpgradeBlockNumber - 1 };
  }
  logger.debug(
    `[PPFS] Fetching strategy implem for ${chain}:${contractAddress} (${JSON.stringify(
      callOptions
    )})`
  );
  const strategyImplem = await callLockProtectedRpc(chain, async (provider) => {
    const contract = new ethers.Contract(
      contractAddress,
      BeefyVaultV6Abi,
      provider
    );
    const [stragegy] =
      callOptions === undefined
        ? await contract.functions.strategy()
        : await contract.functions.strategy(callOptions);
    return stragegy;
  });
  // yield the strategy at the deploy block time
  yield {
    blockNumber: deployBlockNumber,
    datetime: deployBlockDatetime,
    data: {
      implementation: strategyImplem,
    },
  };
  // then yield the rest of our data
  yield* allStrategyEvents;
}
