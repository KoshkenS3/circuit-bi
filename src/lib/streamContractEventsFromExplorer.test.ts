import { explorerLogToERC20TransferEvent } from "./streamContractEventsFromExplorer";

describe("test parsing of explorer log data", () => {
  it("should parse a fantom transfer event", () => {
    const event = {
      address: "0x4c25854f6da3b5848f7b5c71dcb8eee20b292d3e",
      topics: [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x000000000000000000000000d146c053f9ea51199d273c769c4fae2a8e12d7b3",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
      ],
      data: "0x00000000000000000000000000000000000000000000025e00dd7ab9b6665e6e",
      blockNumber: "0x1aa4c53",
      timeStamp: "0x61e17b53",
      gasPrice: "0x3982e23a40",
      gasUsed: "0x3e9ed",
      logIndex: "0xd7",
      transactionHash:
        "0x38a5f8f010d31ce81a075fc04a2c9adf78bd4bd07a08f779bd15c3aed5a718e4",
      transactionIndex: "0x4",
    };
    expect(explorerLogToERC20TransferEvent(event)).toMatchSnapshot();
  });

  it("should parse a bsc transfer event", () => {
    const event = {
      address: "0x13071d48a5fde2735102657e15d1132f92ee8c83",
      topics: [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000009730155214fcad09a4a002f9ca7f9dbf3e706cd7",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
      ],
      data: "0x000000000000000000000000000000000000000000000000248bfe4b25cd809f",
      blockNumber: "0xc0d378",
      timeStamp: "0x6190c567",
      gasPrice: "0x12a05f200",
      gasUsed: "0x3fed4",
      logIndex: "0x21b",
      transactionHash:
        "0xf4ce74f18f980be4b3fbc27f6f2d1774aa119c307ef27927280759559b708d2d",
      transactionIndex: "0x98",
    };
    expect(explorerLogToERC20TransferEvent(event)).toMatchSnapshot();
  });

  it("should parse a metis transfer event", async () => {
    const event = {
      address: "0xea01ca0423acb8476e1d3bae572021c2aa9bd410",
      blockNumber: "0x81916",
      data: "0x00000000000000000000000000000000000000000000000003bc054b9a039800",
      gasPrice: "0x2cb417800",
      gasUsed: "0x6cf87",
      logIndex: "0x4",
      timeStamp: "0x61e81c70",
      topics: [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
        "0x000000000000000000000000010da5ff62b6e45f89fa7b2d8ced5a8b5754ec1b",
        null,
      ],
      transactionHash:
        "0x7647c50775d428ae15f8f332e663903231c32ff735433732d9e32327a7390a84",
      transactionIndex: "0x0",
    };

    expect(explorerLogToERC20TransferEvent(event)).toMatchSnapshot();
  });
});
