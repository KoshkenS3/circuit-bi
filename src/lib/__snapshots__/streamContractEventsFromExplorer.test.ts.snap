// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test parsing of explorer log data should parse a bsc transfer event 1`] = `
Object {
  "blockNumber": 12637048,
  "datetime": 2021-11-14T08:14:31.000Z,
  "from": "0x9730155214fcad09A4a002f9cA7F9dBf3E706cD7",
  "to": "0x0000000000000000000000000000000000000000",
  "value": "2633478005838413983",
}
`;

exports[`test parsing of explorer log data should parse a fantom transfer event 1`] = `
Object {
  "blockNumber": 27937875,
  "datetime": 2022-01-14T13:32:03.000Z,
  "from": "0xd146C053f9ea51199D273C769C4fAE2a8E12d7B3",
  "to": "0x0000000000000000000000000000000000000000",
  "value": "11178789249575889034862",
}
`;

exports[`test parsing of explorer log data should parse a metis transfer event 1`] = `
Object {
  "blockNumber": 530710,
  "datetime": 2022-01-19T14:13:04.000Z,
  "from": "0x0000000000000000000000000000000000000000",
  "to": "0x010dA5FF62B6e45f89FA7B2d8CEd5a8b5754eC1b",
  "value": "269095900000000000",
}
`;
